### ISSUE TITLE
React 17 Support

### ISSUE THREAD

**Issue by pierceglennie** [airbnb/react-dates#2105]
It would be good if react-dates was updated to support React 17. Are there any plans to do this?

**Comment by ljharb** [comment]
Yes; it's blocked on a few thins: enzyme, airbnb-prop-types, and a few others updating to react 17 being the primary ones.

**Comment by ljrodriguez1** [comment]
Do we have an update on this, would be really helpfull to be able to use this on react 17

**Comment by orangecoding** [comment]
Is this project actually still maintained or is it abandon?

**Comment by ljharb** [comment]
@orangecoding yes, it is still maintained.

**Comment by balovbohdan** [comment]
Any updates?

**Comment by joeyparis** [comment]
Can we have an error thrown/logged if we try using `react-dates` on a version >16? Or at least add it to the README? I just spent 3 days trying to debug my themes not working only to realize it's because I'm on react 17.

**Comment by ljharb** [comment]
@joeyparis no, we can't - you're supposed to run `npm ls` yourself and ensure your dep graph is valid. npm 7+, however, will fail to install if your peer deps aren't compatible, so i suggest using that.

**Comment by Brocoleta** [comment]
Please update this!

**Comment by josegl** [comment]
Yes please!

**Comment by m-nathani** [comment]
> @m-nathani no, not until we can test on it, for which see [#2105 (comment)](https://github.com/airbnb/react-dates/issues/2105#issuecomment-825317710)

are these all `enzyme, airbnb-prop-types, and a few others updating to react 17` include something major that would break the functionality ? .. as it appears these are more related to devDependencies.


asking because if we update our apps to `react 17` and perform `npm i --legacy-peer-deps` it would work for us.. however we aren't confident enough that it would break `react-dates` with it... 
Moreover, according to react docs.. `react 17` has no new changes and support gradual updates 
https://reactjs.org/blog/2020/10/20/react-v17.html#no-new-features

**Comment by styrken** [comment]
Would very much like official react 17 support

**Comment by JRebella** [comment]
Is this library being maintained? It's been almost a year we're getting close to React 18 and still no support for 17?

**Comment by ljharb** [comment]
@JRebella yes, it's still being maintained. Unfortunately the passage of time does not help provide me with much time for unpaid work.

Specifically, this project can't support React 17 until its tests pass on React 17, and that's blocked on enzyme, primarily. (which i'm also the only maintainer of).

**Comment by ehartford** [comment]
Why are you marking comments as spam?  I haven't seen any spam here.

**Comment by ljharb** [comment]
@ehartford any comment anywhere on github asking for updates, or effectively saying "me too", is spam. If there was an update, it would be posted on the issue - if there's nothing posted, there's no update, and the comment needlessly notifies everyone subscribed.

**Comment by styrken** [comment]
@ljharb I does not completely agree on the SPAM thesis - It is also a way of showing interest in a particular feature/request, to let maintainers/others know.

**Comment by ljharb** [comment]
@styrken that's what emoji reactions are for.

**Comment by ehartford** [comment]
@ljharb I get it.  But, given that the load is too much for you to handle alone, oughtn't you to put out a call for maintainers for both enzyme and react-dates?  Surely there's enough community interest that you can get some more maintainers on board to help review PRs.  I, for one, would be happy to help.

**Comment by ljharb** [comment]
@ehartford absolutely yes, but these comments aren't helping with that :-)

I'm not aware of many open source projects that *aren't* looking for maintainers and help. The challenge with enzyme is that it requires substantial ramp-up, and while lots have been willing to help, none have stuck around for long. I'd very much welcome any help you or others are willing to provide: https://github.com/enzymejs/enzyme/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22

**Comment by jdwillemse** [comment]
Since enzyme is the bottleneck, would it be an acceptable solution to migrate the tests to https://github.com/testing-library/react-testing-library so this lib can be updated?

**Comment by ljharb** [comment]
@jdwillemse no, for a number of reasons - one is that i'm enzyme's maintainer, so i'm not likely to go for an alternative :-) but also, RTL lacks some of the capabilities required.

**Comment by jdwillemse** [comment]
It seems the community has moved on from Enzyme as you have indicated you're the only maintainer there. Hitching this lib's fate to that of Enzyme's seems likely to cause the death of this one too.

**Comment by ljharb** [comment]
@jdwillemse that’s hostile and unhelpful.

**Comment by gabor-at-reed** [comment]
@ljharb I don't know if you've seen this article: [https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl). The article's author is the maintainer of [@wojtekmaj/enzyme-adapter-react-17](https://www.npmjs.com/package/@wojtekmaj/enzyme-adapter-react-17).

But based on that what @wojtekmaj said, maybe you should reconsider your point if you don't want to see this project dies, which would be a big shame, to be honest.

**Comment by ljharb** [comment]
Yes, I’ve read it, and i think it’s clickbait and wrong in a ton of ways,

A project doesn’t die because a random user of it declares it so.

**Comment by WPaczula** [comment]
React 18 is out and it doesn't look like enzyme will support it (looking at the lack of official support for the previous version). It's not looking well for this package moving forward. I think it's fair to mention that on the readme page.

**Comment by gabor-at-reed** [comment]
@WPaczula Check my comment and the answer to that above. It was exactly about this, but the maintainer (@ljharb ) hid it and marked it as a non-relevant comment (the last two comments above your one). 🤷‍♂️ 🤯 It looks like the maintainer is firmly against that idea where we said they should remove enzyme from this project and use the testing library instead because that is sort of the only major reason why this package is not supporting react v17 and v18 now. But because the maintainer is the maintainer of the enzyme and cannot let that project go, it looks like this project will die as well, which is a shame, in my opinion.

**Comment by ljharb** [comment]
@WPaczula just like react 17, enzyme will eventually support it.

@gabor-at-reed react-testing-library is incapable of replacing enzyme, because it can't test server-rendering. This isn't about my ego or my ability to "let things go". There simply doesn't exist a complete alternative to enzyme.

**Comment by MarcARoberge** [comment]
Maybe you can do a breaking change version of it that drops server-rendering while you find another solution.

**Comment by ljharb** [comment]
@MarcARoberge a) no, you can't drop server-rendering because that's not a thing you can support, it always works, and b) no, sites that don't server-render are broken, and it's FAR better to stay on react 16 forever than to encourage people not to server-render.

**Comment by MarcARoberge** [comment]
@ljharb I am not talking about droping it for life. Just for now, since it is the only issue. Release a version 17.0.0 with a breaking change that SSR is not fully tested and if they want, they can use 16.X version for it.

**Comment by ljharb** [comment]
The not-small effort to do that would be much better spent finishing the enzyme react 17 adapter.

**Comment by BillOTei** [comment]
hope this is moving forward, thanks for your work.

**Comment by medin-n** [comment]
Any update on supporting react 17?

**Comment by Glutnix** [comment]
> The not-small effort to do that would be much better spent finishing the enzyme react 17 adapter.

The Enzyme React 17 Adaptor author Wojciech Maj [declared Enzyme dead in December 2021. ](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl). The plausibility of a React 18 adaptor is also discussed, and they have done the hard work to move away from Enzyme to React Testing Library.

Wojciech is also the maintainer of react-date-picker, this component's main alternative. 

React 16 docs aren't even available, and I have no reason to believe it will recieve LTS or security patches, as Meta/Facebook haven't released a patch for 16 since October 2020.

How long will this project wait for something extremely unlikely to occur to actually take place before it progresses onwards?

**Comment by ljharb** [comment]
@Glutnix that’s an unofficial adapter, and that individual has no place to declare anything dead.

it will wait as long as it takes.

**Comment by tkostas15** [comment]
Can we safely use the package in React 17? Even if we use --legacy-peer-deps, which is necessary. Has anyone tested it?

**Comment by turansky** [comment]
I use it with React 18. [Small correction required](https://github.com/react-dates/react-dates/issues/2105#issuecomment-943447926)

**Comment by ljharb** [comment]
@tkostas15 its highly likely it’ll mostly work. But safely? No, it’s never safe to use software with a version it explicitly claims it’s incompatible with.

**Comment by longlostnick** [comment]
I actually ended up going with react-calendar! For the most part it’s been a dream to work with. It’s not as fully featured, but it’s much more lightweight and easy to customize for the functionality you do want. 

https://github.com/wojtekmaj/react-calendar

**Comment by Stas-Buzunko** [comment]
> I actually ended up going with react-calendar! For the most part it’s been a dream to work with. It’s not as fully featured, but it’s much more lightweight and easy to customize for the functionality you do want.
> 
> https://github.com/wojtekmaj/react-calendar

thanks! 👍

**Comment by Swapnilchavan18** [comment]
@turansky Does it work with react 18? What is the small correction required?

**Comment by turansky** [comment]
@Swapnilchavan18 It works fine with React `18.2.0`. 
[Small correction details](https://github.com/turansky/kfc-plugins/blob/2be905e4f0d4359b32e15031088ff6197135fc81/kfc/kfc-gradle-plugin/src/main/kotlin/io/github/turansky/kfc/gradle/plugin/ReactDatesPlugin.kt#L21)

**Comment by aderiushev** [comment]
total disrespect to the community 👎 
rm -rf

**Comment by suenot** [comment]
npm i react-dates --save
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: my_project@0.1.0
npm ERR! Found: react@18.2.0
npm ERR! node_modules/react
npm ERR!   react@"18.2.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^0.14 || ^15.5.4 || ^16.1.1" from react-dates@21.8.0
npm ERR! node_modules/react-dates
npm ERR!   react-dates@"*" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /root/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2023-04-23T09_03_52_671Z-debug-0.log

**Comment by aerialist7** [comment]
Due to the fact that `react-dates` library is not supported I forked it, then released and publish a new version that includes this fix.
Anyone can find it by the next npm coordinates: [`"@aerialist7/react-dates": "21.8.0-patch.0"`](https://www.npmjs.com/package/@aerialist7/react-dates/v/21.8.0-patch.0)

**Comment by lemmingapex** [comment]
react-dates is a great library and I applaud @ljharb for leading the effort after it's history with airbnb.  react-dates is one of the most premier and standard date picker libraries in the react world:  [~500k weekly downloads](https://www.npmjs.com/package/react-dates) as of today.  Unfortunately, react-dates has not had the recent attention and care that it's users and community deserve and expect.

The last release of react-dates ([21.8.0](https://github.com/react-dates/react-dates/releases/tag/v21.8.0)) was almost 4 years ago (Jan 29, 2020).  The [last commit on master](https://github.com/react-dates/react-dates/commit/b7bad38dcff024d374ee98f972b55b3de9e61289) was almost 2 years ago (Jan 26, 2022).  During that time, [React 17.0.0](https://github.com/facebook/react/blob/main/CHANGELOG.md#1700-october-20-2020) was released three years ago on October 20, 2020 and [React 18.0.0](https://github.com/facebook/react/blob/main/CHANGELOG.md#1800-march-29-2022) was released over 18 months ago on March 29, 2022.

When the build and dependency status in the second line of the readme are failing, users of react-dates don't have to look very far to see that this library could use some maintenance and love.

Nobody here expects that @ljharb accomplish the task of getting react-dates functional with react 17 and react 18 himself.  We understand you are a busy guy and are working on many different projects.  We are here to learn and help, but we need the information to give this great software the love it needs.

Below are a few questions for @ljharb that many here would benefit from knowing:

1.  When do you expect the next release of this library will be and why?
2.  There is great interest in making this library compatible with react 17 and react 18.  Others are willing to help and have tried to keep this project up to date with react: https://github.com/react-dates/react-dates/pull/2235.  This pull request does not appear to be a duplicate of another.  Can you explain the reason it was closed?
3.  What's the path forward with the most utility for the least amount of work to make react-dates functional with react 17 and 18?  Can that be done?  
4.  In your mind, what are the EXACT requirements to get this library functional with react 17 and react 18?  
5.  Would you rather the world have a version of react-dates that can be used by the current version of react and doesn't have the rigorous enzyme testing, or would you rather it continue to suffer code rot?  The last release of react-dates was four years ago. 
6.  If this library is indeed still maintained as you [claim](https://github.com/react-dates/react-dates/issues/2105#issuecomment-993864942) and [claim](https://github.com/react-dates/react-dates/issues/2199#issuecomment-1768948902), then why haven't there been updates to the baseline in 2 years?  Not even the README.md has been updated.
7.  If you are unable to maintain this library on your own, please add some prominent contributors as collaborators and let others continue to find value in this software.  Part of open source software is building a community that can keep moving forward.  This is difficult.  How do we help you?  You can not run the show alone.
8.  If there are alternatives (or forks) to this library that you can recommend, I would suggest updating the appropriate documentation to say so.


Just as everyone else here, I would like to see this project continue to help thousands of people into the future.  I thank you for your time and effort.

**Comment by ljharb** [comment]
1. As soon as I'm able to get enzyme supporting react 17 and 18, then everything I maintain that relies on that will update, including airbnb-prop-types and react-dates. "Why" should be obvious; it's long overdue.
1. This project can't be tested properly on newer versions of react until enzyme supports it, and claiming to support a react version that's not tested is a wildly unacceptable and harmful lie that I won't perpetrate.
1. The path forward rests on enzyme supporting react 17 and 18.
1. See above.
1. I would rather it sit for a CENTURY than claim to support something that's not properly tested, and it's absurd to suggest otherwise.
1. "Maintained" does not require recent updates of any kind, including to the readme - as long as security issues are quickly addressed, it's maintained.
1. The issue, again, is enzyme, something that nobody has chosen to help sufficiently with for almost 7 years. I agree building a community is hard, but the barriers here aren't about inclusivity for other maintainers - the barriers are simply that nobody who's willing to put forth the actual effort required (or the money for others to accomplish it) has materialized.
1. I have no alternatives I can endorse.

**Comment by lemmingapex** [comment]
Thanks for your comments @ljharb.

1.  There is no alternative future for react-dates other than enzyme support for react 17 and react 18 in your mind?  Why is this the case?  
2.  When will enzyme support react 17 and 18?  What if Enzymes never supports react >=17?  That seems just as likely based on history.
3.  Currently users take this code, they hack it together so they can use it with react 17 and react 18.  Isn't than worse than having slightly limited testing?
4.  If nobody has given the effort to enzyme in the last 7 years, why should we continue to use it?

Thanks.

**Comment by ljharb** [comment]
1. There does not currently exist any other equally complete testing solution for React.
1. I can't commit to a date, because as I've mentioned, no people or money have materialized to allow me to make such a commitment. Future likelihoods are unknowable in any direction.
1. Nope, it's not, because I don't have to _support_ that code.
1. It's your choice to use anything; I'd hope that you'd continue to use it because the benefits proper testing provide are often a lot greater than the benefits new features in React provide.

**Comment by lemmingapex** [comment]
You're right, @ljharb, you don't have to support what people do with this library; but ignoring how people actually use this library is worse than ignoring the library all together.  The world changes wether or not we want it to.  I hope that open source finds a sustainable future in a world filled with rotting code.  Maybe in another 7 years you will find the support for enzyme you have been looking for in the past 7 years.  Until then, I expect most will seek an alternative.  🍻

**Comment by ljharb** [comment]
If they have nothing to contribute and aren't satisfied with the free labor of volunteers, they are certainly welcome to seek an alternative.

**Comment by lemmingapex** [comment]
Plenty of volunteers, plenty of support and interest.  That support just doesn't align with how you see the world.  This is bigger than you.  Then again, the stubbornness of the few have always hindered the pursuits of the many.

**Comment by ljharb** [comment]
@lemmingapex that tone convinces me to hide this discussion; this isn't about being stubborn, this is about good engineering practices. Please do seek an alternative.

**Comment by NikhilIluri** [comment]
> Due to the fact that `react-dates` library is not supported I forked it, then released and publish a new version that includes this fix. Anyone can find it by the next npm coordinates: [`"@aerialist7/react-dates": "21.8.0-patch.0"`](https://www.npmjs.com/package/@aerialist7/react-dates/v/21.8.0-patch.0)

@aerialist7 , does this work with react v18 now?

**Comment by turansky** [comment]
> @aerialist7 , does this work with react v18 now?

It works

**Comment by NikhilIluri** [comment]
> > @aerialist7 , does this work with react v18 now?
> 
> It works

still facing the same issue while doing npm i
Found: react@18.2.0
npm ERR! node_modules/react
npm ERR!   react@"18.2.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^0.14 || ^15.5.4 || ^16.1.1" from @aerialist7/react-dates@21.8.0-patch.0
npm ERR! node_modules/@aerialist7/react-dates
npm ERR!   @aerialist7/react-dates@"^21.8.0-patch.0" from the root project
